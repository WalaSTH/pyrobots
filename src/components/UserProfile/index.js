// MUI components
import {
  Avatar,
  Card,
  Divider,
  Grid,
  List,
  ListItem,
  ListItemAvatar,
  ListItemText,
  Typography,
  Badge,
  IconButton,
  Menu,
  MenuItem,
} from "@mui/material";

// MUI icons
import SettingsIcon from "@mui/icons-material/Settings";
import AddPhotoAlternateIcon from "@mui/icons-material/AddPhotoAlternate";
import SportsEsportsIcon from "@mui/icons-material/SportsEsports";
import EmojiEventsIcon from "@mui/icons-material/EmojiEvents";
import PercentIcon from "@mui/icons-material/Percent";
import { useState } from "react";

export default function UserProfile({ username, avatar, stats }) {
  const [anchorEl, setAnchorEl] = useState(null);

  function handleClick(event) {
    setAnchorEl(event.currentTarget);
  }

  function handleClose() {
    setAnchorEl(null);
  }

  const menuOpen = Boolean(anchorEl);

  return (
    <Card
      variant="outlined"
      sx={{
        padding: 3,
        borderRadius: 3,
      }}
    >
      <Grid
        container
        spacing={2}
        sx={{
          display: "flex",
          justifyContent: "center",
        }}
      >
        <Grid
          item
          xs={12}
          sx={{
            display: "flex",
            justifyContent: "right",
            marginBottom: -1.5,
          }}
        >
          <IconButton
            onClick={handleClick}
            sx={{
              marginBottom: -4,
            }}
          >
            <SettingsIcon />
          </IconButton>

          <Menu
            open={menuOpen}
            onClose={handleClose}
            anchorEl={anchorEl}
            anchorOrigin={{
              vertical: "bottom",
              horizontal: "center",
            }}
            transformOrigin={{
              vertical: "top",
              horizontal: "center",
            }}
          >
            <MenuItem onClick={handleClose}>Change user avatar</MenuItem>
            <MenuItem onClick={handleClose}>Change password</MenuItem>
          </Menu>
        </Grid>
        <Grid
          item
          xs={12}
          sx={{
            display: "flex",
            justifyContent: "center",
            marginBottom: -1.5,
          }}
        >
          <IconButton>
            <Badge
              overlap="circular"
              badgeContent={
                <Avatar
                  sx={{
                    width: "1.6rem",
                    height: "1.6rem",
                    backgroundColor: "primary.main",
                  }}
                >
                  <AddPhotoAlternateIcon
                    sx={{
                      width: "1.2rem",
                      height: "1.2rem",
                    }}
                  />
                </Avatar>
              }
            >
              <Avatar
                alt={username}
                src={avatar || "none"}
                sx={{
                  width: "6rem",
                  height: "6rem",
                  fontSize: "2.5rem",
                  border: 2.75,
                  borderColor: "primary.main",
                }}
              />
            </Badge>
          </IconButton>
        </Grid>

        <Grid
          item
          xs={12}
          sx={{
            display: "flex",
            justifyContent: "center",
            marginTop: -0.5,
            marginBottom: -0.5,
          }}
        >
          <Typography sx={{ fontSize: "1.5rem" }}> {username} </Typography>
        </Grid>

        <Grid
          item
          xs={12}
          sx={{
            display: "flex",
            justifyContent: "center",
            marginTop: -0.5,
            marginBottom: -0.5,
          }}
        >
          <Divider variant="middle" sx={{ width: "100%" }} />
        </Grid>

        <Grid item xs={12}>
          <Typography variant="h6" color="primary" padding={1}>
            User stats
          </Typography>

          <Card
            variant="filled"
            sx={{
              backgroundColor: "#f0f7ff",
            }}
          >
            <List>
              <ListItem>
                <ListItemAvatar>
                  <SportsEsportsIcon />
                </ListItemAvatar>
                <ListItemText
                  primary="Games played"
                  secondary={stats.gamesPlayed || "-"}
                />
              </ListItem>

              <ListItem>
                <ListItemAvatar>
                  <EmojiEventsIcon />
                </ListItemAvatar>
                <ListItemText
                  primary="Games won"
                  secondary={stats.gamesWon || "-"}
                />
              </ListItem>

              <ListItem>
                <ListItemAvatar>
                  <PercentIcon />
                </ListItemAvatar>
                <ListItemText
                  primary="Victory rate"
                  secondary={stats.victoryRate ? stats.victoryRate + "%" : "-"}
                />
              </ListItem>
            </List>
          </Card>
        </Grid>
      </Grid>
    </Card>
  );
}
