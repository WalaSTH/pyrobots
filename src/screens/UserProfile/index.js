// import { Container } from "@mui/material";
// import DisplayProfile from "../../components/DisplayProfile";

import { Container } from "@mui/material";
import axios from "axios";
import { useEffect, useState } from "react";
import Snackbar from "../../components/FormsUI/Snackbar";
import ProfileCard from "../../components/ProfileCard";

export default function UserProfile() {
  const [open, setOpen] = useState(false);
  const [body, setBody] = useState("");
  const [severity, setSeverity] = useState("");

  function handleClose(reason) {
    if (reason === "clickaway") return;
    setOpen(false);
  }

  const username = localStorage.getItem("username");
  const avatar = localStorage.getItem("avatar");
  const [stats, setStats] = useState(null);

  async function fetchStats(username) {
    return await axios.get("http://127.0.0.1:8000/users/stats", {
      params: {
        checked_user: username,
      },
    });
  }

  useEffect(() => {
    fetchStats(username)
      .then((res) => {
        setStats(res.data.stats);
      })
      .catch((error) => {
        setSeverity("error");
        if (
          error.response &&
          typeof error.response.data["detail"] != "object"
        ) {
          setBody(error.response.data["detail"]);
        } else {
          setBody("Unknown error");
        }
        setOpen(true);
      });
  }, [username]);

  return (
    <Container
      component="main"
      maxWidth="xs"
      sx={{
        marginTop: 10,
      }}
    >
      <ProfileCard username={username} avatar={avatar} stats={stats} />

      {open && (
        <Snackbar
          data-testid="profileSnackbar"
          open={open}
          body={body}
          severity={severity}
          handleClose={handleClose}
        />
      )}
    </Container>
  );
}
