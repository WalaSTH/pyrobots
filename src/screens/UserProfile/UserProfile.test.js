import "@testing-library/jest-dom/extend-expect";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer, { act } from "react-test-renderer";
import UserProfile from ".";
import axios from "axios";

jest.mock("axios");

describe("UserProfile", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render initial state of the component correctly", async () => {
    await act(async () => render(<UserProfile />));
    // User info
    expect(screen.getByTestId("userAvatar")).toBeInTheDocument();
    expect(screen.getByTestId("userName")).toBeInTheDocument();

    // User settings button
    expect(screen.getByTestId("userSettingsButton")).toBeInTheDocument();

    // User stats
    expect(screen.getByText(/User stats/i)).toBeInTheDocument();
    expect(screen.getByText(/Matches played/i)).toBeInTheDocument();
    expect(screen.getByText(/Matches won/i)).toBeInTheDocument();
    expect(screen.getByText(/Victory rate/i)).toBeInTheDocument();
    expect(screen.queryAllByTestId("statsSkeleton")).toHaveLength(3);

    // Settings menu should be closed
    expect(screen.queryByTestId("userSettingsMenu")).toBeNull();
    expect(screen.queryAllByTestId("menuItem")).toHaveLength(0);
    expect(screen.queryByText(/Change user avatar/i)).toBeNull();
    expect(screen.queryByText(/Change password/i)).toBeNull();
  });

  it("should render stats after succesfully fethching stats", async () => {
    const mockedStats = { played_matches: 1, victories: 1 };

    axios.get.mockResolvedValue({
      data: { stats: mockedStats },
    });

    render(<UserProfile />);

    await waitFor(() => {
      const statsItems = screen.queryAllByTestId("statsItem");
      const statsSkeletons = screen.queryAllByTestId("statsSkeleton");

      expect(statsItems).toHaveLength(3);
      expect(statsSkeletons).toHaveLength(0);

      expect(statsItems[0]).toHaveTextContent(mockedStats.played_matches);
      expect(statsItems[1]).toBeInTheDocument(mockedStats.victories);
      expect(statsItems[2]).toBeInTheDocument(
        mockedStats.victories / mockedStats.played_matches + "%"
      );
    });
  });

  it("should render skeletons and snackbar after erroneously fetching stats", async () => {
    const mockedDetail = "dummy error message";

    axios.get.mockRejectedValue({
      response: { data: { detail: mockedDetail } },
    });

    render(<UserProfile />);

    await waitFor(() => {
      const statsItems = screen.queryAllByTestId("statsItem");
      const statsSkeletons = screen.queryAllByTestId("statsSkeleton");
      const errorSkeleton = screen.getByTestId("profileSnackbar");

      expect(statsItems).toHaveLength(0);
      expect(statsSkeletons).toHaveLength(3);
      expect(errorSkeleton).toBeInTheDocument();
      expect(errorSkeleton).toHaveTextContent(mockedDetail);
    });
  });

  it("should open menu after clicking on the settings icon", async () => {
    const user = userEvent.setup();
    await act(async () => render(<UserProfile />));

    await user.click(screen.getByTestId("userSettingsButton"));

    expect(screen.getByTestId("userSettingsMenu")).toBeInTheDocument();
    expect(screen.queryAllByTestId("menuItem")).toHaveLength(2);
    expect(screen.getByText(/Change avatar/i)).toBeInTheDocument();
    expect(screen.getByText(/Change password/i)).toBeInTheDocument();
  });

  it("should close menu after clicking on any of the options", async () => {
    const user = userEvent.setup();
    await act(async () => render(<UserProfile />));

    await user.click(screen.getByTestId("userSettingsButton"));
    await user.click(screen.getByText(/Change avatar/i));
    await waitFor(() =>
      expect(screen.queryByTestId("userSettingsMenu")).toBeNull()
    );

    await user.click(screen.getByTestId("userSettingsButton"));
    await user.click(screen.getByText(/Change password/i));
    await waitFor(() =>
      expect(screen.queryByTestId("userSettingsMenu")).toBeNull()
    );
  });

  it("matches Snapshot", () => {
    const component = renderer.create(<UserProfile />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
