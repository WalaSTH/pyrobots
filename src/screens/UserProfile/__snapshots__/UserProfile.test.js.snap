// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile matches Snapshot 1`] = `
<main
  className="MuiContainer-root MuiContainer-maxWidthXs css-y2662u-MuiContainer-root"
>
  <div
    className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-9u37zt-MuiPaper-root-MuiCard-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-9qmdtu-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1ec7nnp-MuiGrid-root"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-174qwy3-MuiButtonBase-root-MuiIconButton-root"
          data-testid="userSettingsButton"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="SettingsIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
            />
          </svg>
        </button>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-16t8s8b-MuiGrid-root"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiBadge-root BaseBadge-root css-1c32n2y-MuiBadge-root"
          >
            <div
              className="MuiAvatar-root MuiAvatar-circular css-t2joe1-MuiAvatar-root"
              data-testid="userAvatar"
            >
              <img
                alt={null}
                className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                src="none"
              />
            </div>
            <span
              className="MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightCircular MuiBadge-overlapCircular BaseBadge-badge css-144dnr3-MuiBadge-badge"
            >
              <div
                className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-bgfe38-MuiAvatar-root"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-18bqt7o-MuiSvgIcon-root"
                  data-testid="AddPhotoAlternateIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 7v2.99s-1.99.01-2 0V7h-3s.01-1.99 0-2h3V2h2v3h3v2h-3zm-3 4V8h-3V5H5c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2v-8h-3zM5 19l3-4 2 3 3-4 4 5H5z"
                  />
                </svg>
              </div>
            </span>
          </span>
        </button>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-l59uj0-MuiGrid-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-qjpmex-MuiTypography-root"
          data-testid="userName"
        />
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-l59uj0-MuiGrid-root"
      >
        <hr
          className="MuiDivider-root MuiDivider-middle css-mx8507-MuiDivider-root"
        />
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <h6
          className="MuiTypography-root MuiTypography-h6 css-6uu9lt-MuiTypography-root"
        >
          User stats
        </h6>
        <div
          className="MuiPaper-root MuiPaper-filled MuiPaper-rounded MuiCard-root css-8cn4f-MuiPaper-root-MuiCard-root"
        >
          <ul
            className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
          >
            <li
              className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
              disabled={false}
            >
              <div
                className="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="SportsEsportsIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m21.58 16.09-1.09-7.66C20.21 6.46 18.52 5 16.53 5H7.47C5.48 5 3.79 6.46 3.51 8.43l-1.09 7.66C2.2 17.63 3.39 19 4.94 19c.68 0 1.32-.27 1.8-.75L9 16h6l2.25 2.25c.48.48 1.13.75 1.8.75 1.56 0 2.75-1.37 2.53-2.91zM11 11H9v2H8v-2H6v-1h2V8h1v2h2v1zm4-1c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1zm2 3c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1z"
                  />
                </svg>
              </div>
              <div
                className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
              >
                <span
                  className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                >
                  Matches played
                </span>
                <p
                  className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
                >
                  <span
                    className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                    data-testid="statsSkeleton"
                    style={
                      {
                        "height": undefined,
                        "width": "40%",
                      }
                    }
                  />
                </p>
              </div>
            </li>
            <li
              className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
              disabled={false}
            >
              <div
                className="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="EmojiEventsIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5h-2V3H7v2H5c-1.1 0-2 .9-2 2v1c0 2.55 1.92 4.63 4.39 4.94.63 1.5 1.98 2.63 3.61 2.96V19H7v2h10v-2h-4v-3.1c1.63-.33 2.98-1.46 3.61-2.96C19.08 12.63 21 10.55 21 8V7c0-1.1-.9-2-2-2zM5 8V7h2v3.82C5.84 10.4 5 9.3 5 8zm14 0c0 1.3-.84 2.4-2 2.82V7h2v1z"
                  />
                </svg>
              </div>
              <div
                className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
              >
                <span
                  className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                >
                  Matches won
                </span>
                <p
                  className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
                >
                  <span
                    className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                    data-testid="statsSkeleton"
                    style={
                      {
                        "height": undefined,
                        "width": "40%",
                      }
                    }
                  />
                </p>
              </div>
            </li>
            <li
              className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
              disabled={false}
            >
              <div
                className="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="PercentIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7.5 11C9.43 11 11 9.43 11 7.5S9.43 4 7.5 4 4 5.57 4 7.5 5.57 11 7.5 11zm0-5C8.33 6 9 6.67 9 7.5S8.33 9 7.5 9 6 8.33 6 7.5 6.67 6 7.5 6zM4.0025 18.5832 18.59 3.9955l1.4142 1.4143L5.4167 19.9974zM16.5 13c-1.93 0-3.5 1.57-3.5 3.5s1.57 3.5 3.5 3.5 3.5-1.57 3.5-3.5-1.57-3.5-3.5-3.5zm0 5c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5z"
                  />
                </svg>
              </div>
              <div
                className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
              >
                <span
                  className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
                >
                  Win rate
                </span>
                <p
                  className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
                >
                  <span
                    className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
                    data-testid="statsSkeleton"
                    style={
                      {
                        "height": undefined,
                        "width": "40%",
                      }
                    }
                  />
                </p>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</main>
`;
